File: Difficulty: Easy/Get min at pop/get-min-at-pop.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

// Function to push all the elements into the stack.
stack<int> push(vector<int>& arr) {
    // Create a stack and push all elements from the array into it
    stack<int> s;
    for(int num : arr) {
        s.push(num);
    }
    return s;
}

// Function to print the minimum value in stack each time while popping.
void getMinAtPop(stack<int>& s) {
    // Create a temporary vector to store all elements from the stack
    vector<int> temp;
    while(!s.empty()) {
        temp.push_back(s.top());
        s.pop();
    }
    
    // Push the elements back into the stack in the original order
    for(int i = temp.size() - 1; i >= 0; i--) {
        s.push(temp[i]);
    }
    
    // Now process each pop operation
    while(!s.empty()) {
        // Find the minimum in the current stack by checking all elements
        stack<int> tempStack = s;
        int minVal = tempStack.top();
        
        while(!tempStack.empty()) {
            minVal = min(minVal, tempStack.top());
            tempStack.pop();
        }
        
        // Output the minimum value for this pop operation
        cout << minVal << " ";
        s.pop();  // Pop the element from the stack
    }
}


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        string line;
        getline(cin, line);
        stringstream ss(line);
        vector<int> arr;
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }
        stack<int> s = push(arr);
        getMinAtPop(s);

        cout << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Easy/Get min at pop/get-min-at-pop.js
//{ Driver Code Starts
//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

class Stack{
    constructor(){
        this.arr = [];
        this.top = -1;
    }
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let n = parseInt(readLine());
        let arr = new Array(n);
        let input_ar1 = readLine().split(' ').map(x=>parseInt(x));
        for(let i=0;i<n;i++)
            arr[i] = input_ar1[i];
        let obj = new Solution();
        let mys = obj._push(arr, n);
        obj._getMinAtPop(mys);
        
    }
}
// } Driver Code Ends

// User function Template for JavaScript

class Solution {
    
    /**
     * @param {number[]} arr
     * @param {number} n
     * @returns {Stack}
     */
    // Function to push all the elements into the stack.
    _push(arr, n) {
        // Create a new stack instance
        let stack = new Stack();

        // Iterating over the array elements and pushing them into the stack
        for (let i = 0; i < n; i++) {
            stack.arr.push(arr[i]);
            stack.top++;
        }

        return stack;
    }

    /**
     * @param {Stack} s
     */
    // Function to print the minimum value in the stack each time while popping
    _getMinAtPop(s) {
        let values = [];

        // Pushing all the elements of the stack into a list (array)
        while (s.top !== -1) {
            values.push(s.arr[s.top]);
            s.arr.pop();
            s.top--;
        }

        // Create a new stack to keep track of the minimum elements
        let minStack = new Stack();
        
        // Push the last element into both the original stack and the minStack
        s.arr.push(values[values.length - 1]);
        s.top++;
        minStack.arr.push(values[values.length - 1]);
        minStack.top++;

        // Iterate over the list in reverse order (starting from second last element)
        for (let i = values.length - 2; i >= 0; i--) {
            // Push the current element into the original stack
            s.arr.push(values[i]);
            s.top++;

            // Compare the top elements of both stacks and push the minimum value into minStack
            if (s.arr[s.top] < minStack.arr[minStack.top]) {
                minStack.arr.push(s.arr[s.top]);
                minStack.top++;
            } else {
                minStack.arr.push(minStack.arr[minStack.top]);
                minStack.top++;
            }
        }

        // Print all the elements in minStack one by one
        let result = '';
        while (s.top !== -1) {
            result += minStack.arr[minStack.top] + " ";
            s.arr.pop();
            s.top--;
            minStack.arr.pop();
            minStack.top--;
        }

        console.log(result);
    }
}
